src/string_bytes.cc

```diff
@@ -656,20 +678,108 @@ static void force_ascii(const char* src, char* dst, size_t len) {
 }
 
 
-static size_t hex_encode(const char* src, size_t slen, char* dst, size_t dlen) {
+#define pack(byte) (static_cast<uint8_t>(byte) * 0x01010101u)
+uint32_t nibbles_to_hex(uint32_t nibbles) {
+  const uint32_t ascii09 = nibbles + pack('0');
+  const uint32_t correction = pack('a' - '0' - 10);
+
+  const uint32_t tmp = nibbles + pack(128 - 10);
+  const uint32_t msb = tmp & pack(0x80);
+  const uint32_t mask = msb - (msb >> 7);
+
+  return ascii09 + (mask & correction);
+}
+
+
+#define byte_to_hex(src, dst, i, val)                                          \
+  do {                                                                         \
+    val = static_cast<uint8_t>(src[i]);                                        \
+    dst[i * 2] = hex[val >> 4];                                                \
+    dst[i * 2 + 1] = hex[val & 15];                                            \
+    if (slen == i + 1) return;                                                 \
```

Comment:
Macros with non-local control flow that don't announce it in their name are a special kind of evil.